src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -213,6 +214,55 @@ where
         }
     }
 
+    /// Run the protocol to serve the connection.
+    pub async fn run(mut self) {
+        let mut notice_stream = self.stream.clone();
+
+        loop {
+            let session = self.session.clone();
+
+            let mut process = std::pin::pin!(async {
+                let msg = match self.read_message().await {
+                    Ok(msg) => msg,
+                    Err(e) => {
+                        tracing::error!(error = %e.as_report(), "error when reading message");
+                        return true; // terminate the connection
+                    }
+                };
+                tracing::trace!(?msg, "received message");
+                self.process(msg).await
+            });
+
+            let terminated = if let Some(session) = session {
+                // If a session is present, subscribe and send notices asynchronously
+                // while processing the message.
+                loop {
+                    let next_notice = async {
+                        let notice = session.next_notice().await;
+                        notice_stream.write(&BeMessage::NoticeResponse(&notice)).await.inspect_err(|e| {
+                            tracing::error!(error = %e.as_report(), notice, "failed to send notice");
+                        }).ok();
+                    };
```

Comment:
This is not cancellation-safe.

Reply:
Could you explain more? I guess here no leaking will happen. And if the task is cancelled, it is ok to not send full notice to the user?

Reply:
Changed to spawn a task for sending notices, which is more clear and less error-prone.

Reply:
Changed to not create a `next_notice` future every time when processing a message.